---
title: "Rscript_RNAseq_analysis_per_cell_type"
output: html_document
date: "2023-08-29"
author: "Carli"
update: "2023-08-08"
  chunk_output_type: console
editor_options: 
  chunk_output_type: console
---

If necessary, download the required libraries

```{r}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(version = "3.17")

BiocManager::install("DESeq2")
BiocManager::install("EnhancedVolcano")
BiocManager::install("edgeR")
BiocManager::install("tidyverse")
BiocManager::install("ggrepel")
install.packages("ggplot2")
install.packages("ggfortify")
BiocManager::install("RColorBrewer")
BiocManager::install("pheatmap")
install.packages("reshape")
BiocManager::install("PCAtools")
BiocManager::install("apeglm")
BiocManager::install("annotables")
```

Load the libraries

```{r}
library("DESeq2") #install Bioc::Manager and then DESeq2
library("ggplot2")
library("ggrepel")
library("EnhancedVolcano") #install BioC::Manager and then EnhancedVolcano
library("RColorBrewer")
library("pheatmap")
library("apeglm")
library("edgeR") #install BioC::Manager and then edgeR
library("tidyverse")
library("ggfortify")
library("reshape")
```

Load the metadata and rawcount.csv file

```{r}
setwd("G:/My Drive/RNAseq/Carli_SN+LUVA")
#RNA_project <- readline(prompt="Project code: ") #enter projectcode until second dash
Cell1_name <- readline(prompt="Name cell type 1: ")
Cell2_name <- readline(prompt="Name cell type 2: ")

metadata_Cell1 <- read.csv(file.path("analyse", paste0("Metadata", Cell1_name, ".csv")), row.names = 1)
metadata_Cell2 <- read.csv(file.path("analyse", paste0("Metadata", Cell2_name, ".csv")), row.names = 1)

counts_Cell1 <- read.csv(file.path("analyse", paste0("Rawcounts", Cell1_name, ".csv")), sep=',', row.names = 1)
counts_Cell2 <- read.csv(file.path("analyse", paste0("Rawcounts", Cell2_name, ".csv")), row.names = 1)
```

Check if order of samples is the same in metadata and count table

```{r}
all(rownames(metadata_Cell1) == colnames(counts_Cell1))
all(rownames(metadata_Cell2) == colnames(counts_Cell2))
#match(colnames(counts_Cell1), rownames(metadata_Cell1)) #see how metadata needs to be reordered to match the rawcounts

#use outcome of match to reorder the metadata
#idx <- match(colnames(counts_Cell1), roawnames(metadata_Cell1)) 
#reordered_metadata_Cell1 <- metadata_Cell1[idx,]

```

Create DESEq2 object

```{r}
dds_Cell1 <- DESeqDataSetFromMatrix(countData = counts_Cell1, 
                                  colData = metadata_Cell1, 
                                  design = ~ batch + condition) 
dds_Cell1$condition <- factor(dds_Cell1$condition, levels = c("co", "mono"))

dds_Cell2 <- DESeqDataSetFromMatrix(countData = counts_Cell2, 
                                    colData = metadata_Cell2, 
                                    design = ~ batch + condition) #mono is first in metadata so seen as base level
dds_Cell2$condition <- factor(dds_Cell2$condition, levels = c("co", "mono"))

```

Prefiltering - only keep rows that have at least 10 reads in total

```{r}
keep_Cell1 <- rowSums(counts(dds_Cell1)) >= 10
dds_Cell1 <- dds_Cell1[keep_Cell1,]

keep_Cell2 <- rowSums(counts(dds_Cell2)) >= 10
dds_Cell2 <- dds_Cell2[keep_Cell2,]
```

Estimate library size factors #needed before normalized counts extraction

```{r}
dds_Cell1 <- estimateSizeFactors(dds_Cell1) #calculates normalized counts 
#every sample has a size factor for which it should be normalized
#sizeFactors(dds_Cell1) #show size factors
dds_Cell2 <- estimateSizeFactors(dds_Cell2)
```

Everything count normalization to linear modelling + fitting raw counts to negative binomial model

```{r}
dds_Cell1 <- DESeq(dds_Cell1)
dds_Cell2 <- DESeq(dds_Cell2)

# Now you can access results with the desired contrast name
resultsNames(dds_Cell2)

#dds_Cell2$condition = relevel(dds_Cell2$condition, ref = "co")

#dds_Cell2 <- estimateSizeFactors(dds_Cell2)
#dds_Cell2 <- estimateDispersions(dds_Cell2)
#dds_Cell2 <- nbinomWaldTest(dds_Cell2)

# Now you can access results with the desired contrast name
#resultsNames(dds_Cell2)

# Add condition
resGenotype1 <- results(dds_Cell1, contrast = c("condition", "co", "mono"))
head(resGenotype1)

resGenotype2 <- results(dds_Cell2, contrast = c("condition", "co", "mono")) #sets co as base level - meaning logfoldchanges show changes of co on mono
head(resGenotype2)

#plot the dispersion plots; should match the slope of the red line
plotDispEsts(dds_Cell1)
resultsNames(dds_Cell1)

plotDispEsts(dds_Cell2)
resultsNames(dds_Cell2)

meancounts1 <- apply(counts_Cell1[,1:3],1,mean) #calculate the mean for each gene (=each row)
variancecounts1 <- apply(counts_Cell1[,1:10],1,var)
df1 <- data.frame(meancounts1, variancecounts1)
ggplot(df1) +
    geom_point(aes(x=meancounts1, y=variancecounts1)) +
    scale_y_log10()+
    scale_x_log10()+
    xlab("Mean counts per gene") +
    ylab("Variance per gene")

meancounts2 <- apply(counts_Cell2[,1:3],1,mean) #calculate the mean for each gene (=each row)
variancecounts2 <- apply(counts_Cell2[,1:10],1,var)
df2 <- data.frame(meancounts2, variancecounts2)
ggplot(df2) +
    geom_point(aes(x=meancounts2, y=variancecounts2)) +
    scale_y_log10()+
    scale_x_log10()+
    xlab("Mean counts per gene") +
    ylab("Variance per gene")
```

Extract the normalized counts

```{r}
normalized_Cell1_counts <- counts(dds_Cell1, normalized = TRUE) #extracted counts normalized by estimateSIzeFactors()
write.csv(normalized_Cell1_counts, file.path("./analyse/", paste0("Normalized Counts", Cell1_name, ".csv"))) 

normalized_Cell2_counts <- counts(dds_Cell2, normalized = TRUE)
write.csv(normalized_Cell2_counts, file.path("./analyse/", paste0("Normalized Counts", Cell2_name, ".csv"))) 

```

Log transformation #waarom een log transformatie voor invoering PCA? #waarom niet de extracted normalized counts? #zijn deze data uit dds ook genormaliseerd?

```{r}
vsd_Cell1 <- vst(dds_Cell1, blind = TRUE) #performs log transformation of the data across the mean
vsd_mat_Cell1 <- assay(vsd_Cell1) # extract the matrix of the transformed counts
vsd_cor_Cell1 <- cor(vsd_mat_Cell1) #compute the correlation values between samples
pheatmap(vsd_cor_Cell1, annotation = select(metadata_Cell1, condition))

#vsd_mat_Cell1 <- assay(vsd_Cell1) # extract the matrix of the transformed counts
#vsd_cor_Cell1 <- cor(vsd_mat_Cell1) #compute the correlation values between samples
#pheatmap(vsd_cor_Cell1, annotation = select(metadata_Cell1, condition))

vsd_Cell2 <- vst(dds_Cell2, blind = TRUE)
vsd_mat_Cell2 <- assay(vsd_Cell2) # extract the matrix of the transformed counts
vsd_cor_Cell2 <- cor(vsd_mat_Cell2) #compute the correlation values between samples
pheatmap(vsd_cor_Cell2, annotation = select(metadata_Cell2, condition))

```

PCA #principal component analysis

```{r}
PCA_Cell1 <- plotPCA(vsd_Cell1, intgroup = c("condition"))
PCA_Cell1 + geom_text_repel(aes_string(x = "PC1", y = "PC2", label = "name"), color = "black", size = 2)

#change labels
label_names <- c("exp1", "exp2", "exp3", "exp4", "exp5", "exp1", "exp2", "exp3", "exp4", "exp5") # Replace ... with your label names
vsd_Cell1$label <- label_names
PCA_Cell1 <- plotPCA(vsd_Cell1, intgroup = c("condition"))
PCA_Cell1 + geom_text_repel(aes_string(x = "PC1", y = "PC2", label = "label"), color = "black", size = 2)

print(PCA_Cell1)

PCA_Cell2 <- plotPCA(vsd_Cell2, intgroup = c("condition"))
#plotPCA(vsd_Cell2, intgroup = "condition")
PCA_Cell2 + geom_text_repel(aes_string(x = "PC1", y = "PC2", label = "name"), color = "black", size = 2)
```

Extract results

```{r}
Cell1_results <- results(dds_Cell1,
                       name = "condition_mono_vs_co",
                       alpha = 0.5)
summary(Cell1_results)
head(Cell1_results)

resGenotype1 <- results(dds_Cell1, contrast = c("condition", "co", "mono"))
head(resGenotype1)

#Cell1_results <- results(dds_Cell1, contrast = c("condition", "co", "mono"),
#                      name = "condition_mono_vs_co",
#                       alpha = 0.5)
#summary(Cell1_results)

Cell2_results <- results(dds_Cell2,
                       name = "condition_mono_vs_co",
                       alpha = 0.5) 
summary(Cell2_results)

#Cell2_results <- results(dds_Cell2, contrast = c("condition", "co", "mono"),
#                       name = "condition_mono_vs_co",
#                       alpha = 0.5) 
#summary(Cell2_results)
#sets co as base level - meaning logfoldchanges show changes of co on mono
```

Log fold shrinkage #outliers eruit? is dat echt nodig? - allen voor MA plot

```{r}
LFC_Cell1_results <- lfcShrink(dds_Cell1, 
                            coef = "condition_mono_vs_co", 
                            res = Cell1_results,
                            type = "apeglm")
LFC_Cell1_results$log2FoldChange <- -1 * LFC_Cell1_results$log2FoldChange
summary(LFC_Cell1_results)
head(LFC_Cell1_results)
#can add lfcthreshold: lfcThreshold = 0.3
#name of wanted comparison model e.g. grcm38 for mice
#library(annotables) finds genenames to ensmbl codes

#generate more likely, lower log2foldchange estimates, without chanigng the number of differentially expressed genes that are returned
LFC_Cell2_results <- lfcShrink(dds_Cell2,
                             coef = "condition_mono_vs_co", 
                             res = Cell2_results)
LFC_Cell2_results$log2FoldChange <- -1 * LFC_Cell2_results$log2FoldChange
summary(Cell2_results)
head(LFC_Cell2_results)
```

MA plot #fold change vs mean expression

```{r}
plotMA(LFC_Cell1_results, ylim = c(-20, 20)) #edit
DESeq2::plotMA(LFC_Cell1_results, ylim = c(-15, 15))

plotMA(LFC_Cell2_results, ylim = c(-20, 20)) #edit
DESeq2::plotMA(LFC_Cell2_results, ylim = c(-5, 15))
```

*VOLCANO PLOT* 1. obtain full results table and order by significant results

```{r}
#first re-extract results with correct comparison (co vs mono)
Cell1_results <- results(dds_Cell1, contrast = c("condition", "co", "mono"),
                       name = "condition_mono_vs_co",
                       alpha = 0.5)
summary(Cell1_results)
head(Cell1_results)

Cell2_results <- results(dds_Cell2, contrast = c("condition", "co", "mono"),
                       name = "condition_mono_vs_co",
                       alpha = 0.5)
summary(Cell1_results)
head(Cell1_results)

Cell1_res_df <- data.frame(Cell1_results)
Cell2_res_df <- data.frame(Cell2_results)

Cell1_arranged <- arrange(Cell1_res_df, padj) 
Cell2_arranged <- arrange(Cell2_res_df, padj)
```

2.  save results table -\> convert gene ID to symbol for sig genes -\> import

```{r}
write.csv(Cell1_arranged, file.path("analyse", paste0("ID", Cell1_name, ".csv"))) 
write.csv(Cell2_arranged, file.path("analyse", paste0("ID", Cell2_name, ".csv"))) 

Cell1_sig_ID <- arrange(Cell1_res_df, padj) 

Cell1_sig_ID <- read.csv(file.path("analyse", paste0("ID", Cell1_name, ".csv")), sep = ",")
Cell2_sig_ID <- read.csv(file.path("analyse", paste0("ID", Cell2_name, ".csv")), sep = ";")
```

3.  plot using enhanced volcano

```{r}
EnhancedVolcano(Cell1_sig_ID,
                lab = Cell1_sig_ID$X,
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-5, 5), #edit
                ylim = c(0, 100), #edit
                title = "Effect of coculture on LUVA",
                pCutoff = 0.05,
                FCcutoff = 1.5, #1
                pointSize = 2.0,
                labSize = 4.0,
                col=c('grey', 'green4', 'blue', 'red3'),
                colAlpha = 1)

EnhancedVolcano(Cell2_sig_ID,
                lab = (Cell2_sig_ID[,1]),
                x = 'log2FoldChange',
                y = 'padj',
                xlim = c(-10, 10),
                ylim = c(0, 80),
                title = 'Effect of coculture on LUVA',
                pCutoff = 0.05,
                FCcutoff = 1.5,
                pointSize = 2.0,
                labSize = 4.0,
                col=c('grey', 'green4', 'blue', 'red3'),
                colAlpha = 1)
```

*SNP asthma genes - Zaïd*
```{r}
#import list of SNP genes
SNP_genes <- read.csv(file.path("analyse", "SNP_genes.csv"))

# Subset Cell1_sig_ID to keep only the rows with all genes
Cell1_sig_ID_filtered <- Cell1_sig_ID[Cell1_sig_ID$X %in% SNP_genes$Genes, ]
Cell2_sig_ID_filtered <- Cell2_sig_ID[Cell2_sig_ID$X %in% SNP_genes$Genes, ]

# Write the
write.csv(Cell1_sig_ID_filtered, file.path("analyse", paste0("SNP_filtered", Cell1_name, ".csv"))) 
write.csv(Cell2_sig_ID_filtered, file.path("analyse", paste0("SNP_filtered", Cell2_name, ".csv"))) 
```

*Th2 asthma genes*
```{r}
#import list of SNP genes
Th2_genes <- read.csv(file.path("analyse", "Th2_genes.csv"))
Th2Go_genes <- read.csv(file.path("analyse", "Th2GO_genes.csv"))

# Subset Cell1_sig_ID to keep only the rows with all genes
Cell1_sig_ID_Thfiltered <- Cell1_sig_ID[Cell1_sig_ID$X %in% Th2_genes$Genes, ]
Cell2_sig_ID_Thfiltered <- Cell2_sig_ID[Cell2_sig_ID$X %in% Th2_genes$Genes, ]
Cell1_sig_ID_ThGOfiltered <- Cell1_sig_ID[Cell1_sig_ID$X %in% Th2Go_genes$Genes, ]
Cell2_sig_ID_ThGOfiltered <- Cell2_sig_ID[Cell2_sig_ID$X %in% Th2Go_genes$Genes, ]

# Write the
write.csv(Cell1_sig_ID_Thfiltered, file.path("analyse", paste0("Th2_filtered", Cell1_name, ".csv"))) 
write.csv(Cell2_sig_ID_Thfiltered, file.path("analyse", paste0("Th2_filtered", Cell2_name, ".csv"))) 
write.csv(Cell1_sig_ID_ThGOfiltered, file.path("analyse", paste0("Th2GO_filtered", Cell1_name, ".csv"))) 
write.csv(Cell2_sig_ID_ThGOfiltered, file.path("analyse", paste0("Th2GO_filtered", Cell2_name, ".csv"))) 
```

*Heatmap - Th2*
```{r}
#export all normalized counts
write.csv(Cell2_res_df, file.path("analyse", paste0("ResultsDEseq", Cell2_name, ".csv")))

#manually link it to the genes needed for Manuscript GUilherme
#inner_join(GenesGuilherme, normalized_Cell2_counts, by = Gene)

#Cell2_Guilherme_genes <- read.csv(file.path("analyse", "Genes Guilherme_counts_noSN10.csv"), row.names=1)
#Cell2_Guilherme_genes <- read.csv(file.path("analyse", "GenesGuilherme_pseudo_zscores-no10.csv"), row.names=1)
Cell2_Th2all_genes <- read.csv(file.path("analyse", "Th2_genes_heatmap.csv"), row.names=1)

pheatmap(Cell2_Th2all_genes, 
         color = colorRampPalette(c("blue", "white", "red"))(50), 
         cluster_cols = F,
         cluster_rows = F,
         show_rownames = T, 
         annotation = select(metadata_Cell2, condition),
         scale = "row",
         height = 1000) 

```

*Heatmap - mast cell differentiation - Guilherme*
```{r}
#export all normalized counts
write.csv(Cell2_res_df, file.path("analyse", paste0("ResultsDEseq", Cell2_name, ".csv")))

#manually link it to the genes needed for Manuscript GUilherme
#inner_join(GenesGuilherme, normalized_Cell2_counts, by = Gene)

#Cell2_Guilherme_genes <- read.csv(file.path("analyse", "Genes Guilherme_counts_noSN10.csv"), row.names=1)
#Cell2_Guilherme_genes <- read.csv(file.path("analyse", "GenesGuilherme_pseudo_zscores-no10.csv"), row.names=1)
Cell2_Guilherme_genes <- read.csv(file.path("analyse", "GenesGuilherme_zscores-noSN10-noCGRP-final.csv"), row.names=1)

pheatmap(Cell2_Guilherme_genes, 
         color = colorRampPalette(c("blue", "white", "red"))(50), 
         cluster_cols = F,
         cluster_rows = F,
         show_rownames = T, 
         annotation = select(metadata_Cell2, condition),
         scale = "row",
         height = 1000) 

```

*HEATMAP* Of all genes

```{r}
Cell1_res_sig <- subset(Cell1_res_df, padj < 0.05) %>%
  arrange(padj)

Cell1_sig_norm_counts <- normalized_Cell1_counts[rownames(Cell1_res_sig),]
#display.brewer.all()
heat.colors <- brewer.pal (n = 6, name = "RdYlBu")

pheatmap(Cell1_sig_norm_counts, 
         color = colorRampPalette(c("blue", "white", "red"))(50), 
         cluster_rows = T,
         show_rownames = F, 
         annotation = select(metadata_Cell1, condition),
         scale = "row")

###

Cell2_res_sig <- subset(Cell2_res_df, padj < 0.05) %>%
  arrange(padj)

Cell2_sig_norm_counts <- normalized_Cell2_counts[rownames(Cell2_res_sig),]
#display.brewer.all()
heat.colors <- brewer.pal (n = 6, name = "RdYlBu")

pheatmap(Cell2_sig_norm_counts, 
         color = colorRampPalette(c("blue", "white", "red"))(50), 
         cluster_rows = T,
         show_rownames = F, 
         annotation = select(metadata_Cell2, condition),
         scale = "row")
```

of top50 genes (padj)

```{r}
Cell1_padj_top50_genes <- head(Cell1_res_sig, 50)
Cell1_padj_top50_sig_norm_counts <- merge(normalized_Cell1_counts, Cell1_padj_top50_genes, by = 0)
values_Cell1_padj_top50_sig_norm_counts <- Cell1_padj_top50_sig_norm_counts[,2:9]

Cell1_padj_top50_genes_ID <- head(Cell1_sig_ID, 50)
row.names(values_Cell1_padj_top50_sig_norm_counts) <- row.names(Cell1_padj_top50_genes) #geneID = names
#row.names(values_Cell1_padj_top50_sig_norm_counts) <- row.names(Cell1_padj_top50_genes_ID) #geneID = numbers

pheatmap(values_Cell1_padj_top50_sig_norm_counts, 
         color = colorRampPalette(c("blue", "white", "red"))(50), 
         cluster_rows = T,
         show_rownames = T, 
         annotation = select(metadata_Cell1, condition),
         scale = "row") 

###

Cell2_padj_top50_genes <- head(Cell2_res_sig, 50)
Cell2_padj_top50_sig_norm_counts <- merge(normalized_Cell2_counts, Cell2_padj_top50_genes, by = 0)
values_Cell2_padj_top50_sig_norm_counts <- Cell2_padj_top50_sig_norm_counts[,2:9]

Cell2_padj_top50_genes_ID <- head(Cell2_sig_ID, 50)
row.names(values_Cell2_padj_top50_sig_norm_counts) <- row.names(Cell2_padj_top50_genes) #geneID = names
#row.names(values_Cell2_padj_top50_sig_norm_counts) <- row.names(Cell2_padj_top50_genes_ID) #geneID = numbers

pheatmap(values_Cell2_padj_top50_sig_norm_counts, 
         color = colorRampPalette(c("blue", "white", "red"))(50), 
         clustering_distance_cols = "correlation",
         cluster_cols = T,
         cluster_rows = T,
         show_rownames = T, 
         annotation = select(metadata_Cell2, condition),
         scale = "row",
         height = 1000) 
```
